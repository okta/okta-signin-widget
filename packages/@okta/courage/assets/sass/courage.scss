// Import styles from Nim
@import 'nim/nim';

@import 'fonts/all';
@import 'helpers/all';
@import 'icons/all';
@import 'typography/all';
@import 'from-saasure';

// ------ Components ------
@import 'o-forms/all';
@import 'data-lists/all';
@import 'instruction-list/all';
@import 'dropdown/all';
@import 'o-wizard/all';
@import 'o-file-uploader/all';
@import 'vendor/selectize';
@import 'vendor/selectize_overrides';
@import 'sidebar/all';
@import 'button/all';
@import 'infobox/all';
@import 'tabs/all';
@import 'modal/all';
// ------ Components ------
