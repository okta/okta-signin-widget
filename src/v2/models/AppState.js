/*!
 * Copyright (c) 2019, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { _, Model } from 'okta';
import Logger from 'util/Logger';

export default Model.extend({

  local: {
    introspectSuccess: 'object', // only set during introspection
    introspectError: 'object', // only set during introspection

    factor: 'object',      // optional
    user: 'object',        // optional
    currentFormName: 'string',
    idx: 'object',
    remediations: 'array',
    __previousResponse: 'object'
  },

  derived: {
    authenticatorProfile: {
      // While we're moving toward `authenticator` platform, but still
      // need to support `factor` for certain period.
      // Could remove `factor` after it's deprecated completely.
      deps: ['currentAuthenticator', 'authenticatorEnrollment', 'factor'],
      fn (currentAuthenticator = {}, authenticatorEnrollment = {}, factor = {}) {
        return currentAuthenticator.profile
          || authenticatorEnrollment.profile
          || factor.profile
          || {};
      },
    },
    authenticatorType: {
      deps: ['currentAuthenticator', 'authenticatorEnrollment', 'factor'],
      fn (currentAuthenticator = {}, authenticatorEnrollment = {}, factor = {}) {
        return currentAuthenticator.type
          || authenticatorEnrollment.type
          || factor.factorType
          || '';
      },
    },
    showSignoutLink: {
      deps: ['idx'],
      fn: function (idx = {}) {
        const invalidSignOutSteps = ['IDENTIFY', 'ENROLL', 'SUCCESS'];
        // hide signout for IDENTIFY, ENROLL & SUCCESS step
        return idx.actions
          && _.isFunction(idx.actions.cancel) && !invalidSignOutSteps.includes(idx.context.step);
      },
    },
  },

  hasRemediationObject (formName) {
    return this.get('idx').neededToProceed.find((remediation) => remediation.name === formName);
  },

  getActionByPath (actionPath) {
    const paths = actionPath.split('.');
    let targetObject;
    if (paths.length === 1) {
      targetObject = this.get('idx').actions;
    } else {
      targetObject = this.get(paths.shift());
    }
    // Limitation
    // At the time of writting, action only lives in first level of state objects.
    const actionName = paths.shift();
    if (targetObject && _.isFunction(targetObject[actionName])) {
      return targetObject[actionName];
    } else {
      return null;
    }
  },

  getCurrentViewState () {
    const currentFormName = this.get('currentFormName');

    let currentViewState;
    if (!_.isEmpty(this.get('remediations'))) {
      currentViewState = this.get('remediations').filter(r => r.name === currentFormName)[0];
    }

    if (!currentViewState) {
      if (currentFormName) {
        Logger.warn(`Cannot find view state for form ${currentFormName}. Fall back to terminal state.`);
      }

      currentViewState = this.get('terminal');
    }

    return currentViewState;
  },

  setIonResponse (idxResp) {
    // Don't re-render view if new response is same as last.
    // Usually happening at polling and pipeline doesn't proceed to next step.
    // expiresAt will be different for each response, hence compare objects without that property
    if (_.isEqual(_.omit(idxResp.rawIdxState, 'expiresAt'), _.omit(this.get('__previousResponse'), 'expiresAt'))) {
      return;
    }

    // `currentFormName` is default to first form of remediation object or nothing.
    idxResp.currentFormName = null;

    if (idxResp.neededToProceed && idxResp.rawIdxState.remediation) {
      idxResp.currentFormName = idxResp.rawIdxState.remediation.value[0].name;
    }

    if (idxResp.rawIdxState.success) {
      idxResp.currentFormName = idxResp.rawIdxState.success.name;
    }

    // default terminal state for fall back
    //TODO: This is a FailSafe, it needs to be removed later by this Jira: OKTA-300044
    if (idxResp.context.terminal && _.isEmpty(idxResp.context.terminal.value)) {
      idxResp.terminal = {
        name: 'terminal',
        value: [],
        uiSchema: [],
      };
    }

    // default terminal state for fall back
    if (idxResp.context.messages) {
      idxResp.terminal = {
        name: 'terminal',
        value: idxResp.context.messages.value.length ? idxResp.context.messages.value : [],
        uiSchema: [],
      };
    }
    //clear appState before setting new values
    this.clear({silent: true});
    // set new app state properties
    this.set(idxResp);
    // reset idx
    this.set('idx', idxResp);
    this.set('__previousResponse', idxResp.rawIdxState);

    // broadcast idxResponseUpdated to re-render the view
    this.trigger('idxResponseUpdated', idxResp);
  }

});

// Keep track of stateMachine with this special model. Similar to Appstate.js
