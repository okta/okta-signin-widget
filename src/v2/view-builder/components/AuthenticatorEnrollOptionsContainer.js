/*!
 * Copyright (c) 2020, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */
import { loc, View, Collection } from '@okta/courage';
import { FORMS as RemediationForms } from '../../ion/RemediationConstants';
import AuthenticatorEnrollOptions from './AuthenticatorEnrollOptions';
import remindMeLater from './RemindMeLaterButton';
import skipAll from './SkipOptionalEnrollmentButton';

export default View.extend({

  className: 'authenticator-enroll-list-container',

  initialize: function() {
    const authenticatorsWithGracePeriod = [];
    const authenticators = [];

    this.options.optionItems.forEach((authenticator) => {
      if (authenticator.relatesTo?.gracePeriod?.expiry
        && new Date(authenticator.relatesTo?.gracePeriod?.expiry).getTime() > Date.now()) {
        authenticatorsWithGracePeriod.push(authenticator);
      } else {
        authenticators.push(authenticator);
      }
    });

    this.hasSkipRemediation = !!this.options.appState.hasRemediationObject(RemediationForms?.SKIP);
    this.hasOnlyOptional = this.hasSkipRemediation && authenticatorsWithGracePeriod?.length === 0;
    this.hasOnlyGracePeriod =  this.hasSkipRemediation &&
      this.options.optionItems?.length === authenticatorsWithGracePeriod?.length;

    if (authenticators.length) {
      this.add(new AuthenticatorEnrollOptions({
        ...this.options,
        collection: new Collection(authenticators),
        optionItems: authenticators,
        listTitle: this.hasOnlyOptional
          ? loc('oie.setup.optional.short', 'login') : loc('oie.setup.required.now', 'login'),
      }));
    }
    if (authenticatorsWithGracePeriod.length) {
      this.add(new AuthenticatorEnrollOptions({
        ...this.options,
        collection: new Collection(authenticatorsWithGracePeriod),
        optionItems: authenticatorsWithGracePeriod,
        listTitle: loc('oie.setup.required.soon', 'login'),
        listSubtitle: loc('oie.setup.required.soon.description', 'login'),
      }));
    }

    if (this.hasOnlyGracePeriod) {
      this.add(remindMeLater);
    } else if (this.hasOnlyOptional) {
      this.add(skipAll);
    }
  },
});
