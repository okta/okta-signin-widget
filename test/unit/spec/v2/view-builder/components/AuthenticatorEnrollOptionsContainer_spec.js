import { Collection } from '@okta/courage';
import AppState from 'v2/models/AppState';
import Settings from 'models/Settings';
import AuthenticatorEnrollOptionsContainer from 'v2/view-builder/components/AuthenticatorEnrollOptionsContainer';
import authenticatorEnrollResponse from '../../../../../../playground/mocks/data/idp/idx/authenticator-enroll-grace-period.json';

describe('v2/view-builder/components/AuthenticatorEnrollOptionsContainer', function() {
  let testContext;

  beforeEach(function() {
    testContext = {};
    testContext.init = (
      authenticators,
      skipRemediation,
    ) => {
      const currentViewState = {
        name: 'select-authenticator-enroll',
      };
      const appState = new AppState({}, {});
      appState.set('user', authenticatorEnrollResponse.user.value);
      appState.set(
        'remediation',
        authenticatorEnrollResponse.remediation.value
      );
      appState.set('idx', {neededToProceed: []});

      const settings = new Settings({ baseUrl: 'http://localhost:3000' });

      jest.spyOn(Date, 'now').mockReturnValue(new Date('12/30/2023').getTime());
      jest.spyOn(appState, 'hasRemediationObject').mockReturnValue(skipRemediation);

      testContext.view = new AuthenticatorEnrollOptionsContainer({
        currentViewState,
        appState,
        settings,
        collection: new Collection(authenticators),
        optionItems: authenticators,
        name: 'authenticator',
      });
      testContext.view.render();
    };
  });

  it('renders required now list for all expired grace periods', function() {
    const authenticators = [
      {
        label: 'Okta Phone',
        value: {
          id: 'aid568g3mXgtID0X1SLH',
        },
        relatesTo: {
          label: 'Okta Phone',
          id: 'phone-enroll-id-123',
          type: 'phone',
          key: 'phone_number',
          authenticatorId: 'aid568g3mXgtID0X1SLH',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': '2019-12-17T05:00:00.000Z'
          },
          allowedFor: 'any',
        },
        authenticatorKey: 'phone_number',
      },
      {
        label: 'Security Key or Biometric Authenticator',
        value: {
          id: 'aidtheidkwh282hv8g3',
        },
        relatesTo: {
          displayName: 'Security Key or Biometric Authenticator (FIDO2)',
          type: 'security_key',
          key: 'webauthn',
          id: 'webauthn-enroll-id-123',
          authenticatorId: 'aidtheidkwh282hv8g3',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': '2019-12-17T05:00:00.000Z'
          },
          allowedFor: 'any'
        },
        authenticatorKey: 'webauthn',
      },
    ];
    testContext.init(authenticators, false);
    expect(testContext.view.el).toMatchSnapshot();
    expect(testContext.view.$('.authenticator-list-title').length).toBe(1);
    expect(testContext.view.$('.authenticator-list-title').text()).toBe('Required now');
    expect(testContext.view.$('.authenticator-row').length).toBe(2);
    expect(testContext.view.$('.authenticator-usage-text').length).toBe(2);
    expect(testContext.view.$('.authenticator-grace-period-text-container').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-required-description').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-expiry-date').length).toBe(0);
    expect(testContext.view.$('.skip-all').length).toBe(0);
  });

  it('renders required soon list for all non-expired grace periods', function() {
    const authenticators = [
      {
        label: 'Okta Phone',
        value: {
          id: 'aid568g3mXgtID0X1SLH',
        },
        relatesTo: {
          label: 'Okta Phone',
          id: 'phone-enroll-id-123',
          type: 'phone',
          key: 'phone_number',
          authenticatorId: 'aid568g3mXgtID0X1SLH',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': '2023-12-31T05:00:00.000Z'
          },
          allowedFor: 'any',
        },
        authenticatorKey: 'phone_number',
      },
      {
        label: 'Security Key or Biometric Authenticator',
        value: {
          id: 'aidtheidkwh282hv8g3',
        },
        relatesTo: {
          displayName: 'Security Key or Biometric Authenticator (FIDO2)',
          type: 'security_key',
          key: 'webauthn',
          id: 'webauthn-enroll-id-123',
          authenticatorId: 'aidtheidkwh282hv8g3',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': '2023-12-30T23:00:00.000Z'
          },
          allowedFor: 'any'
        },
        authenticatorKey: 'webauthn',
      },
    ];
    testContext.init(authenticators, true);
    expect(testContext.view.el).toMatchSnapshot();
    expect(testContext.view.$('.authenticator-list-title').length).toBe(1);
    expect(testContext.view.$('.authenticator-list-title').text()).toBe('Required soon');
    expect(testContext.view.$('.authenticator-row').length).toBe(2);
    expect(testContext.view.$('.authenticator-usage-text').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-text-container').length).toBe(2);
    expect(testContext.view.$('.authenticator-grace-period-required-description').length).toBe(2);
    expect(testContext.view.$('.authenticator-grace-period-expiry-date').length).toBe(2);
    expect(testContext.view.$('.skip-all').length).toBe(1);
    expect(testContext.view.$('.skip-all').text()).toBe('Remind me later');
  });

  it('renders required now and required soon for expired and non-expired grace periods', function() {
    const authenticators = [
      {
        label: 'Okta Phone',
        value: {
          id: 'aid568g3mXgtID0X1SLH',
        },
        relatesTo: {
          label: 'Okta Phone',
          id: 'phone-enroll-id-123',
          type: 'phone',
          key: 'phone_number',
          authenticatorId: 'aid568g3mXgtID0X1SLH',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': '2020-12-17T05:00:00.000Z'
          },
          allowedFor: 'any',
        },
        authenticatorKey: 'phone_number',
      },
      {
        label: 'Security Key or Biometric Authenticator',
        value: {
          id: 'aidtheidkwh282hv8g3',
        },
        relatesTo: {
          displayName: 'Security Key or Biometric Authenticator (FIDO2)',
          type: 'security_key',
          key: 'webauthn',
          id: 'webauthn-enroll-id-123',
          authenticatorId: 'aidtheidkwh282hv8g3',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': '2025-12-17T05:00:00.000Z'
          },
          allowedFor: 'any'
        },
        authenticatorKey: 'webauthn',
      },
    ];
    testContext.init(authenticators, false);
    expect(testContext.view.el).toMatchSnapshot();
    expect(testContext.view.$('.authenticator-list-title').length).toBe(2);
    expect(testContext.view.$('.authenticator-list-title')[0].textContent).toBe('Required now');
    expect(testContext.view.$('.authenticator-list-title')[1].textContent).toBe('Required soon');
    expect(testContext.view.$('.authenticator-row').length).toBe(2);
    expect(testContext.view.$('.authenticator-usage-text').length).toBe(1);
    expect(testContext.view.$('.authenticator-grace-period-text-container').length).toBe(1);
    expect(testContext.view.$('.authenticator-grace-period-required-description').length).toBe(1);
    expect(testContext.view.$('.authenticator-grace-period-expiry-date').length).toBe(1);
    expect(testContext.view.$('.skip-all').length).toBe(0);
  });

  it('renders required now list for no grace periods', function() {
    const authenticators = [
      {
        label: 'Okta Phone',
        value: {
          id: 'aid568g3mXgtID0X1SLH',
        },
        relatesTo: {
          label: 'Okta Phone',
          id: 'phone-enroll-id-123',
          type: 'phone',
          key: 'phone_number',
          authenticatorId: 'aid568g3mXgtID0X1SLH',
          allowedFor: 'any',
        },
        authenticatorKey: 'phone_number',
      },
      {
        label: 'Security Key or Biometric Authenticator',
        value: {
          id: 'aidtheidkwh282hv8g3',
        },
        relatesTo: {
          displayName: 'Security Key or Biometric Authenticator (FIDO2)',
          type: 'security_key',
          key: 'webauthn',
          id: 'webauthn-enroll-id-123',
          authenticatorId: 'aidtheidkwh282hv8g3',
          allowedFor: 'any'
        },
        authenticatorKey: 'webauthn',
      },
    ];
    testContext.init(authenticators, false);
    expect(testContext.view.el).toMatchSnapshot();
    expect(testContext.view.$('.authenticator-list-title').length).toBe(1);
    expect(testContext.view.$('.authenticator-list-title')[0].textContent).toBe('Required now');
    expect(testContext.view.$('.authenticator-row').length).toBe(2);
    expect(testContext.view.$('.authenticator-usage-text').length).toBe(2);
    expect(testContext.view.$('.authenticator-grace-period-text-container').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-required-description').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-expiry-date').length).toBe(0);
    expect(testContext.view.$('.skip-all').length).toBe(0);
  });

  it('renders required now list for badly formatted grace period dates', function() {
    const authenticators = [
      {
        label: 'Okta Phone',
        value: {
          id: 'aid568g3mXgtID0X1SLH',
        },
        relatesTo: {
          label: 'Okta Phone',
          id: 'phone-enroll-id-123',
          type: 'phone',
          key: 'phone_number',
          authenticatorId: 'aid568g3mXgtID0X1SLH',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': 'aksjdfhkawef'
          },
          allowedFor: 'any',
        },
        authenticatorKey: 'phone_number',
      },
      {
        label: 'Security Key or Biometric Authenticator',
        value: {
          id: 'aidtheidkwh282hv8g3',
        },
        relatesTo: {
          displayName: 'Security Key or Biometric Authenticator (FIDO2)',
          type: 'security_key',
          key: 'webauthn',
          id: 'webauthn-enroll-id-123',
          authenticatorId: 'aidtheidkwh282hv8g3',
          gracePeriod: {
            'id': 'gpe4hiasrPJX4zwZY789',
            'expiry': ''
          },
          allowedFor: 'any'
        },
        authenticatorKey: 'webauthn',
      },
    ];
    testContext.init(authenticators, false);
    expect(testContext.view.el).toMatchSnapshot();
    expect(testContext.view.$('.authenticator-list-title').length).toBe(1);
    expect(testContext.view.$('.authenticator-list-title')[0].textContent).toBe('Required now');
    expect(testContext.view.$('.authenticator-row').length).toBe(2);
    expect(testContext.view.$('.authenticator-usage-text').length).toBe(2);
    expect(testContext.view.$('.authenticator-grace-period-text-container').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-required-description').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-expiry-date').length).toBe(0);
    expect(testContext.view.$('.skip-all').length).toBe(0);
  });

  it('renders optional authenticators', function() {
    const authenticators = [
      {
        label: 'Okta Phone',
        value: {
          id: 'aid568g3mXgtID0X1SLH',
        },
        relatesTo: {
          label: 'Okta Phone',
          id: 'phone-enroll-id-123',
          type: 'phone',
          key: 'phone_number',
          authenticatorId: 'aid568g3mXgtID0X1SLH',
          allowedFor: 'any',
        },
        authenticatorKey: 'phone_number',
      },
      {
        label: 'Security Key or Biometric Authenticator',
        value: {
          id: 'aidtheidkwh282hv8g3',
        },
        relatesTo: {
          displayName: 'Security Key or Biometric Authenticator (FIDO2)',
          type: 'security_key',
          key: 'webauthn',
          id: 'webauthn-enroll-id-123',
          authenticatorId: 'aidtheidkwh282hv8g3',
          allowedFor: 'any'
        },
        authenticatorKey: 'webauthn',
      },
    ];
    testContext.init(authenticators, true);
    expect(testContext.view.el).toMatchSnapshot();
    expect(testContext.view.$('.authenticator-list-title').length).toBe(1);
    expect(testContext.view.$('.authenticator-list-title')[0].textContent).toBe('Optional');
    expect(testContext.view.$('.authenticator-row').length).toBe(2);
    expect(testContext.view.$('.authenticator-usage-text').length).toBe(2);
    expect(testContext.view.$('.authenticator-grace-period-text-container').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-required-description').length).toBe(0);
    expect(testContext.view.$('.authenticator-grace-period-expiry-date').length).toBe(0);
    expect(testContext.view.$('.skip-all').length).toBe(1);
    expect(testContext.view.$('.skip-all').text()).toBe('Continue');
  });
});
