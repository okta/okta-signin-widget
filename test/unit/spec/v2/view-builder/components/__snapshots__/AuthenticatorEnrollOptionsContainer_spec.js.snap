// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v2/view-builder/components/AuthenticatorEnrollOptionsContainer renders optional authenticators 1`] = `
<div
  class="authenticator-enroll-list-container"
>
  <div
    class="authenticator-enroll-list authenticator-list"
  >
    <div
      class="list-content"
    >
      <div
        class="authenticator-list-title"
      >
        Optional
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Okta Phone
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Security Key or Biometric Authenticator
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <a
    class="button-primary button skip-all link-button"
    data-se="button"
    href="#"
  >
    Continue
  </a>
</div>
`;

exports[`v2/view-builder/components/AuthenticatorEnrollOptionsContainer renders required now and required soon for expired and non-expired grace periods 1`] = `
<div
  class="authenticator-enroll-list-container"
>
  <div
    class="authenticator-enroll-list authenticator-list"
  >
    <div
      class="list-content"
    >
      <div
        class="authenticator-list-title"
      >
        Required now
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Okta Phone
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="authenticator-enroll-list authenticator-list"
  >
    <div
      class="list-content"
    >
      <div
        class="authenticator-list-title"
      >
        Required soon
      </div>
      <p
        class="authenticator-list-subtitle"
      >
        When the deadline passes, you wonâ€™t be able to sign in until you complete setup.
      </p>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Security Key or Biometric Authenticator
          </h3>
          <span
            class="authenticator-grace-period-required-icon"
          />
          <div
            class="authenticator-grace-period-text-container"
          >
            <p
              class="authenticator-grace-period-required-description"
            >
              Required in 718 days
            </p>
            <p
              class="authenticator-grace-period-expiry-date"
            >
              12/17/2025, 05:00 AM UTC
            </p>
          </div>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`v2/view-builder/components/AuthenticatorEnrollOptionsContainer renders required now list for all expired grace periods 1`] = `
<div
  class="authenticator-enroll-list-container"
>
  <div
    class="authenticator-enroll-list authenticator-list"
  >
    <div
      class="list-content"
    >
      <div
        class="authenticator-list-title"
      >
        Required now
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Okta Phone
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Security Key or Biometric Authenticator
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`v2/view-builder/components/AuthenticatorEnrollOptionsContainer renders required now list for badly formatted grace period dates 1`] = `
<div
  class="authenticator-enroll-list-container"
>
  <div
    class="authenticator-enroll-list authenticator-list"
  >
    <div
      class="list-content"
    >
      <div
        class="authenticator-list-title"
      >
        Required now
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Okta Phone
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Security Key or Biometric Authenticator
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`v2/view-builder/components/AuthenticatorEnrollOptionsContainer renders required now list for no grace periods 1`] = `
<div
  class="authenticator-enroll-list-container"
>
  <div
    class="authenticator-enroll-list authenticator-list"
  >
    <div
      class="list-content"
    >
      <div
        class="authenticator-list-title"
      >
        Required now
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Okta Phone
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Security Key or Biometric Authenticator
          </h3>
          <p
            class="authenticator-usage-text"
          >
            Used for access or recovery
          </p>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`v2/view-builder/components/AuthenticatorEnrollOptionsContainer renders required soon list for all non-expired grace periods 1`] = `
<div
  class="authenticator-enroll-list-container"
>
  <div
    class="authenticator-enroll-list authenticator-list"
  >
    <div
      class="list-content"
    >
      <div
        class="authenticator-list-title"
      >
        Required soon
      </div>
      <p
        class="authenticator-list-subtitle"
      >
        When the deadline passes, you wonâ€™t be able to sign in until you complete setup.
      </p>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Okta Phone
          </h3>
          <span
            class="authenticator-grace-period-required-icon"
          />
          <div
            class="authenticator-grace-period-text-container"
          >
            <p
              class="authenticator-grace-period-required-description"
            >
              Required in 1 day
            </p>
            <p
              class="authenticator-grace-period-expiry-date"
            >
              12/31/2023, 05:00 AM UTC
            </p>
          </div>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
      <div
        class="authenticator-row clearfix"
      >
        <div
          class="authenticator-icon-container"
        >
          <div
            aria-label="Authenticator logo"
            class="factor-icon authenticator-icon "
            role="img"
          />
        </div>
        <div
          class="authenticator-description"
        >
          <h3
            class="authenticator-label no-translate"
          >
            Security Key or Biometric Authenticator
          </h3>
          <span
            class="authenticator-grace-period-required-icon"
          />
          <div
            class="authenticator-grace-period-text-container"
          >
            <p
              class="authenticator-grace-period-required-description"
            >
              Required today
            </p>
            <p
              class="authenticator-grace-period-expiry-date"
            >
              12/30/2023, 11:00 PM UTC
            </p>
          </div>
          <div
            class="authenticator-button"
          >
            <a
              class="button select-factor link-button"
              data-se="button"
              href="#"
            >
              Set up
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
  <a
    class="button-primary button skip-all link-button"
    data-se="button"
    href="#"
  >
    Remind me later
  </a>
</div>
`;
